library(shiny)
library(DT)
library(shinyWidgets)
library(adhesiomeR)
library(shinycssloaders)

data(adhesins_df)

shinyUI(navbarPage(title = "adhesiomeR",
                   id = "adhesiomer",
                   tabPanel("Introduction",
                            sidebarPanel(width = 5,
                                         includeMarkdown("intro.md")),
                            mainPanel(width = 7,
                                      dataTableOutput("adhesins"))),
                   tabPanel("Input & settings",
                            sidebarPanel(width = 6,
                                         h5(tags$b("Step 1.")),
                                         fileInput("seq_file", 
                                                   "Upload your genome file(s) in a FASTA format. You can process up to 100 files at once.",
                                                   multiple = TRUE
                                         ),
                                         h5(tags$b("Select number of threads used when running BLAST:")),
                                         numericInput("n_threads", 
                                                      label = NULL,
                                                      value = 1, 
                                                      min = 1, 
                                                      max = parallel::detectCores(logical = FALSE), 
                                                      width = "100px"),
                                         actionButton("blast", "Run BLAST!",
                                                      style="color: #fff; background-color: #337ab7; border-color: #2e6da4"),
                                         HTML("<hr>"),
                                         h5(tags$b("Step 2.")),
                                         pickerInput("systems",
                                                     "Select systems you want to analyse:",
                                                     choices = unique(adhesins_df[["System"]]),
                                                     selected = unique(adhesins_df[["System"]]),
                                                     multiple = TRUE,
                                                     pickerOptions(actionsBox = TRUE,
                                                                   selectedTextFormat = 'count',
                                                                   countSelectedText = '{0} systems selected')),
                                         HTML("<hr>"),
                                         h5(tags$b("Select thresholds for gene presence: ")),
                                         fluidRow(column(4,
                                                         numericInput("identity", "Identity [%]",
                                                                      value = 75,
                                                                      min = 1,
                                                                      max = 100,
                                                                      step = 1,
                                                                      width = "100px"))),
                                         #          column(4,
                                         #                 numericInput("evalue", "E-value",
                                         #                              value = prettyNum(1e-100, scientific = TRUE),
                                         #                              min = 0,
                                         #                              max = 1,
                                         #                              width = "100px"))),
                                         HTML("<hr>"),
                                         h5(tags$b("Customise plot colors: ")),
                                         fluidRow(column(4,
                                                         textInput("presence_col", "Presence",
                                                                   value = "#e00e00",
                                                                   width = "100px")),
                                                  column(4,
                                                         textInput("absence_col", "Absence",
                                                                      value = "#85c1ff",
                                                                      width = "100px"))),
                                         actionButton("filtering_button", "Apply changes!",
                                                      style="color: #fff; background-color: #337ab7; border-color: #2e6da4"),
                            ),
                            mainPanel(width = 6,
                                      dataTableOutput("input_tab"))
                   ),
                   
                   tabPanel("Search results",
                            value = "blast_res",
                            dataTableOutput("blast_res")),
                   tabPanel("Gene presence",
                            value = "all_genes",
                            tabsetPanel(
                              tabPanel("Plot",
                                        checkboxInput("all_genes_hide_missing",
                                                      "Hide genes not found in any genome.",
                                                      value = FALSE),
                                        plotOutput("presence_plot")),
                              tabPanel("Table",
                            dataTableOutput("presence_table")))),
                   tabPanel("System presence",
                            value = "summary_plot",
                            tabsetPanel(
                              tabPanel("Plot",
                                       checkboxInput("systems_summary_hide_missing",
                                                     "Hide systems not found in any genome.",
                                                     value = FALSE),
                                       plotOutput("systems_summary_plot")),
                              tabPanel("Table",
                                       dataTableOutput("systems_summary_table"))
                            )),
                   tabPanel("Gene/system presence",
                            value = "systems",
                            checkboxInput("systems_hide_missing",
                                          "Hide systems not found in any genome.",
                                          value = FALSE),
                            uiOutput("systems_plots")),
                   tabPanel("Clustering",
                            value = "clustering",
                            radioButtons("pathotype",
                                         "Select coloring scheme:",
                                          choices = c("Detailed: aEPEC, tEPEC, EHEC, ETEC, EAEC, EIEC, DAEC, STEC, UPEC, NMEC, Nonpathogenic, NA (unknown)" = "detailed",
                                                      "Intestinal: InPEC (intestinal), ExPEC (extraintestinal), Nonpathogenic, NA (unknown)" = "intestinal"),
                                         width = 900),
                            checkboxInput("clustering_labels",
                                          "Show labels with filenames of analysed genomes",
                                          value = TRUE,
                                          width = 900),
                            withSpinner(plotOutput("clustering_plot",
                                       height = "800px"))),
                   tabPanel("Report",
                            value = "report",
                            checkboxGroupInput("elements",
                                               width = 900,
                                               "Select which elements you want to include in the report:",
                                               choices = c("Table with percents of genes found in each system" = "summary_table",
                                                           "Plot with percents of genes found in each system" = "summary_plot",
                                                           "Table with gene presence/absence" = "presence_table",
                                                           "Plot with gene presence/absence" = "presence_plot"),
                                               selected = c("summary_table", "summary_plot", "presence_table", "presence_plot")),
                            h5(tags$b("Options: ")),
                            checkboxInput("report_hide_genes",
                                          width = 900,
                                          "Hide genes that were not found in any genome.",
                                          value = FALSE),
                            checkboxInput("report_hide_systems",
                                          width = 900,
                                          "Hide systems that were not found in any genome.",
                                          value = FALSE),
                            downloadButton("download",
                                           "Download report"))
                   
                   
))
